### PR TITLE
Replace monad-control with unliftio-core

### LINKED ISSUE #7: Replace `MonadBaseControl` with `MonadUnliftIO`
Hi, would you accept a patch that replaces the dependency on `monad-control` with `unliftio-core`? (`MonadBaseControl m` -> `MonadUnliftIO m`).

The docs for `immortal` read:

> The monadic «state» (captured by the `MonadBaseControl` instance) will be preserved if the computation terminates normally, and reset when the exception is thrown, so be cautious when `m` is stateful. It is completely safe, however, to instantiate m with something like `ReaderT conf IO` to pass configuration to the new thread.

which is the problem `MonadUnliftIO` is meant to solve.

### PR DESCRIPTION
**Pull request by mitchellwrosen**
Fixes #7

### CODE REVIEW

File: src/Control/Immortal.hs [hunk 1]
```diff
@@ -124,24 +114,24 @@ threadId (Thread pid _ _) = pid
...
+  :: MonadUnliftIO m
   => Thread
   -> (Either SomeException () -> m ())
   -> m ()
   -> m ()
-onUnexpectedFinish (Thread _ stopRef _) cb a = do
-  r <- try a
-  expected <- liftBase $ readIORef stopRef
+onUnexpectedFinish (Thread _ stopRef _) cb a = withRunInIO $ \run -> do
+  r <- try $ run a
```
Comment (feuerbach) [inline src/Control/Immortal.hs]: Wouldn't it be cleaner to depend on unlift and use the unlifted functions?

I genuinely don't know; how much here could be simplified with unlift?

File: tests/test.hs [hunk 2]
```diff
@@ -54,28 +53,6 @@ main = defaultMain $ testGroup "Tests"
...
-          countToFive tv
-          liftIO delay
-          error "bah!"
-      pid <- flip evalStateT 0 $ Immortal.create $ const computation
-      (do
-        threadDelay (5*10^5)
-        v <- atomically $ readTVar tv
-        v @?= 0)
-        `finally` Immortal.stop pid
-
```
Comment (feuerbach) [inline tests/test.hs]: Why are the tests removed?

**Review by UnkindPartition — COMMENTED** [review]
Thanks; see my comments. Also, this change fails the CI — could you fix that?

File: src/Control/Immortal.hs (hunk 1 above)
Comment (mitchellwrosen) [reply src/Control/Immortal.hs]: Depending on `unlift` would provide a (rougly) equivalent `try`, `mask`, etc. but -

1. It was easy enough to write with `unliftio-core`
2. `unliftio` would bring in a few extra dependencies like `directory`, `filepath`, `process`, `unix`, so I decided to try to get away without it
3. The semantics would technically change a bit as the unlifted functions ignore asynchronous exceptions by default. The backgrounded thread could then be brought down by a `killThread` without being `mortalized` first, for example, so the logic would have to be tidied up a bit.

File: tests/test.hs (hunk 2 above)
Comment (mitchellwrosen) [reply tests/test.hs]: They no longer compile - `StateT IO` is not an instance of `MonadUnliftIO`. That's part of the reason of switching to `unliftio` - weird stateful `monad-control` instances are ruled out.

**Comment by mitchellwrosen** [comment]
Fixed CI :)

**Comment by UnkindPartition** [comment]
Thanks! Will release shortly.

**Pull request merged**